Update files_antivirus (#3592)
owncloud/docs: modules/admin_manual/pages/configuration/server/virus-scanner-support.adoc

== ClamAV

=== ClamAV Feature List

* Operates on all major operating systems, including _Windows_, _Linux_, and _macOS_.
* Detects all forms of malware including _Trojan horses_, _viruses_, and _worms_.
* Scans _compressed files_, _executables_, _image files_, _Flash_, _PDF_, as well as many others.

What's more, ClamAV's Freshclam daemon automatically updates its malware signature database at scheduled intervals. 

=== ClamAV Integration Into ownCloud

ownCloud integrates with antivirus tools by connecting to them via:

> * A URL and port

Author: This was already there, not my changes - seems that more things need to be revisioned 😢

> * A URL and port

Author: fixed

* A socket
* Streaming the data from the command-line via a pipe with a configured executable

> In the case of ClamAV, ownCloud's Antivirus extension sends files as streams to a ClamAV service (which can be on the same ownCloud server or another server within the same network) which scans them and returns a result to stdout.

Author: fixed

> NOTE: Individual chunks are *not* scanned. The whole file is scanned when it is moved or saved to the final location.

Author: This is the original text, just made it part of the flow of the item so it can be identified where it belongs to.

> NOTE: Individual chunks are *not* scanned. The whole file is scanned when it is moved or saved to the final location.

Author: > Not sure about that. Afaik we do/can do both: scan chunks and scan the assembled file.

Can you calirify and/or propose a better text?

> The information is then parsed, or an exit code is evaluated if no result is available to determine the response from the scan. Based on ownCloud's evaluation of the response (or exit code), an appropriate action is taken, such as recording a log message or deleting the file.

Author: fixed

=== Configuring ClamAV within ownCloud

IMPORTANT: If the app is enabled but either not or incorrectly configured, it will *strictly reject all uploads* for the whole instance!

ClamAV can be configured in following two ways:

. xref:configure-clamav-using-the-av-configuration-panel[By using the Antivirus Configuration panel]
. xref:configure-clamav-using-occ[By using the `occ config:app` command set.]

==== Change Log Level Temporary

> Once ClamAV is installed, select menu:Settings[General (Admin)] and, in the "*Log*" section, temporarily set btn:[Log level] to "_Everything (fatal issues, errors, warnings, info, debug)_".

Author: I guess that the original writer a) had only a "normal" install in his mind and no docker and b) wanted to highlight that a increased loglevel will help finding issues early because a non working av will block all uploads. Means from my pov at least a valid point. Ok if docker turns back the loglevel when restarting automatically., nothing to extra mention in my pov.

> Once ClamAV is installed, select menu:Settings[General (Admin)] and, in the "*Log*" section, temporarily set btn:[Log level] to "_Everything (fatal issues, errors, warnings, info, debug)_".

Author: Leaving as it is, can be improved in a later PR

IMPORTANT: You must run `freshclam` at least once for ClamAV to generate the socket.

image:apps/files_antivirus/antivirus-daemon-socket.png[image]

. Set btn:[Mode] to "*Daemon (Socket)*". ownCloud should detect your `clamd` socket and fill in the "*Socket*" field. This is the `LocalSocket` option in `clamd.conf`. 
+
You can run `ss` (a utility to investigate sockets) to verify it, as in the example below:
+
[source,console]
----
sudo ss -a | grep -iq clamav && echo "ClamAV is running"
----
+
[TIP]
====
If you don't have `ss` installed, you may have `netstat` installed. If so, you can run the following to check if ClamAV is running:
[source,console]
----
netstat -a | grep -q clam && echo "ClamAV is running"
----
====

. When infected files were found during a background scan, you have the choice of either:
+
* Logging any alerts without deleting the files 
* Immediately deleting infected files

===== Daemon (Different Server)

> In _Daemon_ mode, ClamAV runs on a different server. This is a good option for ownCloud servers with high volumes of file uploads. 

Author: fixed

|

|`av_infected_action`
|The action to take when infected files were found during a background scan. +
It can be set to one of `only_log` and `delete`.
|`only_log`

|`av_max_file_size`
|The maximum file size limit; +
`-1` means no limit.
|`-1`

|`av_mode`
|The Anti Virus binary operating mode. +
It can be set to one of `executable`, `daemon`, and `socket`.
|`executable`

|`av_path`
|The path to the `clamscan` executable.
|`/usr/bin/clamscan`

|`av_port`
|The port number of the Antivirus server. +
Allowed values are `1 - 65535`.
|

|`av_socket`
|The name of ClamAV's UNIX socket file.
|`/var/run/clamav/clamd.ctl`

|`av_stream_max_length`
|The maximum stream length that ClamAV will accept in bytes (*).
|`26214400`
|===

(*) The `Stream Length` value sets the number of bytes to read in one pass and defaults to 26214400 bytes (twenty six megabytes). This value should be no larger than the PHP `memory_limit` settings or physical memory if `memory_limit` is set to -1 (no limit).

==== Configuration Modes

ClamAV runs in one of three modes: 

* xref:daemon-socket[Daemon (Socket)]
* xref:daemon[Daemon]
* xref:executable[Executable]

===== Daemon (Socket, Same Server)

> In _Daemon (Socket)_ mode, ClamAV runs in the background on the same server as the ownCloud installation, or the socket can be made available via a share mount. When there is no activity, `clamd` places a minimal load on your system. Consider that high CPU usage which can occur when users upload large volumes of files.

Reviewer: ```suggestion
In _Daemon (Socket)_ mode, ClamAV runs in the background on the same server as the ownCloud installation, or the socket can be made available via a share mount. When there is no activity, `clamd` places a minimal load on your system. Consider that high CPU usage can occur when users upload large volumes of files.
```

IMPORTANT: You must run `freshclam` at least once for ClamAV to generate the socket.

image:apps/files_antivirus/antivirus-daemon-socket.png[image]

. Set btn:[Mode] to "*Daemon (Socket)*". ownCloud should detect your `clamd` socket and fill in the "*Socket*" field. This is the `LocalSocket` option in `clamd.conf`. 
+
You can run `ss` (a utility to investigate sockets) to verify it, as in the example below:
+
[source,console]
----
sudo ss -a | grep -iq clamav && echo "ClamAV is running"
----
+
[TIP]
====
If you don't have `ss` installed, you may have `netstat` installed. If so, you can run the following to check if ClamAV is running:
[source,console]
----
netstat -a | grep -q clam && echo "ClamAV is running"
----
====

. When infected files were found during a background scan, you have the choice of either:
+
* Logging any alerts without deleting the files 
* Immediately deleting infected files

===== Daemon (Different Server)

> In _Daemon_ mode, ClamAV runs on a different server. This is a good option to reduce load on the ownCloud servers when high network bandwidth is available and many concurrent uploads.

Reviewer: ```suggestion
In _Daemon_ mode, ClamAV runs on a different server. This is a good option to reduce load on the ownCloud servers when high network bandwidth is available and many concurrent uploads happen.
```

image:apps/files_antivirus/antivirus-daemon.png[image]

. Set btn:[Mode] to "*Daemon*". 
. Set btn:[Host] to the hostname or IP address of the remote server running ClamAV, and set btn:[Port] to the server's port number.
+
TIP: The port number is the value of `TCPSocket` in `/etc/clamav/clamd.conf`.

===== Executable

In _Executable_ mode, ClamAV runs on the same server as the ownCloud installation, with the `clamscan` command running only when a file is uploaded. 

TIP: `clamscan` can respond slower and may not always be reliable for on-demand usage; it is better to use one of the daemon modes.

image:apps/files_antivirus/antivirus-executable.png[image, width=80%]

The image shows a command line option `--allmatch=yes` (continue scanning within the file after finding a match) which is not necessary to be set and just used here for demonstration purposes of the field.

> NOTE: Starting with ownCloud Anti-Virus version 1.0.0, the path to `clamscan` and the command line options are set via a config.php entry and are read-only in the user interface. Please see the xref:configuration/server/config_apps_sample_php_parameters.adoc[config.php parameters for apps] for more details.

Reviewer: ```suggestion
NOTE: Starting with ownCloud Anti-Virus version 1.0.0, the path to `clamscan` and the command line options are set via a config.php entry and are read-only in the user interface. Refer to the xref:configuration/server/config_apps_sample_php_parameters.adoc[config.php parameters for apps] for more details.
```

. The IP address needs to be set in the general configuration above.

. Now set the mode to 'c-icap with ClamAV':
+
[source,console,subs="attributes+"]
----
{occ-command-example-prefix} occ config:app:set files_antivirus \
    av_request_service --value="avscan"
----

. Next, set the response header:
+
[source,console,subs="attributes+"]
----
{occ-command-example-prefix} config:app:set files_antivirus \
    av_response_header --value="X-Infection-Found"
----

### Run with Kaspersky Anti-Virus (KAV)

. Install Kaspersky ScanEngine for Kubernetes.7z and follow the instructions.

. Set KAV as virus scanner with the value `req`:
+
[source,console,subs="attributes+"]
----
{occ-command-example-prefix} config:app:set files_antivirus \
    av_request_service --value="req"
----

. Set the respective response header:
+
[source,console,subs="attributes+"]
----
{occ-command-example-prefix} config:app:set files_antivirus \
    av_response_header --value="X-Virus-ID"
----
+
NOTE: Older versions of KAV did not send back the virus/infection name in an ICAP header. Version 2.0.0 of KAV allows you to configure the header to transport the virus. By default no header is sent.

=== Run with McAfee

> // Which product exactly?

Author: Will add, thanks for notifying

. Follow this procedure to configure ownCloud for the Kaspersky ScanEngine.

. To use KAV, set the mode to `req` either from the Web interface or via command line:
+
[source,console,subs="attributes+"]
----
{occ-command-example-prefix} config:app:set files_antivirus \
    av_request_service --value="req"
----

. Set the respective response header:
+
[source,console,subs="attributes+"]
----
{occ-command-example-prefix} config:app:set files_antivirus \
    av_response_header --value="X-Virus-ID"
----
+
NOTE: Older versions of KAV did not send back the virus/infection name in an ICAP header. Starting with version 2.0.0 of KAV allows you to configure the header to transport the virus. By default no header is sent.

=== Run with McAfee

Note, McAfee version 7.8.2 and up provide ICAP support. Follow this procedure to configure ownCloud for the McAfee virus scanner.

. Install McAfee based on their instructions.

. To use McAfee, set the mode to `wwreqmod` either from the Web interface or via command line:
+
[source,console,subs="attributes+"]
----
{occ-command-example-prefix} occ config:app:set files_antivirus \
    av_request_service --value="wwreqmod"
----
+

> NOTE: McAfee does not offer predefined response headers.

Author: After reading a lot, it seems you can manually define headers, but nothing that you find out easily. We can recheck later again.
